models/issue_dependency.go

```diff
@@ -0,0 +1,137 @@
+	IssueID      int64     `xorm:"UNIQUE(watch) NOT NULL"`
+	DependencyID int64     `xorm:"UNIQUE(watch) NOT NULL"`
+	Created      time.Time `xorm:"-"`
+	CreatedUnix  int64     `xorm:"INDEX created"`
+	Updated      time.Time `xorm:"-"`
+	UpdatedUnix  int64     `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (exists bool, err error) {
+	sess := x.NewSession()
+
+	// Check if it aleready exists
+	exists, err = issueDepExists(x, issue.ID, dep.ID)
+	if err != nil {
+		return exists, err
+	}
+
+	// If it not exists, create it, otherwise show an error message
+	if !exists {
+		newIssueDependency := &IssueDependency{
+			UserID:       user.ID,
+			IssueID:      issue.ID,
+			DependencyID: dep.ID,
+		}
+
+		if _, err := x.Insert(newIssueDependency); err != nil {
+			return exists, err
+		}
+
+		// Add comment referencing the new dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+			return exists, err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+			return exists, err
+		}
+	}
+	return exists, nil
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+
+	// Check if it exists
+	var exists bool
+	if exists, err = issueDepExists(x, issue.ID, dep.ID); err != nil {
+		return err
+	}
+
+	// If it exists, remove it, otherwise show an error message
+	if exists {
+
+		var issueDepToDelete IssueDependency
+
+		switch depType {
+		case DependencyTypeBlockedBy:
+			issueDepToDelete = IssueDependency{IssueID: issue.ID, DependencyID: dep.ID}
+		case DependencyTypeBlocking:
+			issueDepToDelete = IssueDependency{IssueID: dep.ID, DependencyID: issue.ID}
+		default:
+			return
+		}
+
+		if _, err := x.Delete(&issueDepToDelete); err != nil {
+			return err
+		}
+
+		// Add comment referencing the removed dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, false); err != nil {
+			return err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (exists bool, err error) {
+
+	exists, err = e.Where("(issue_id = ? AND dependency_id = ?) OR (issue_id = ? AND dependency_id = ?)", issueID, depID, depID, issueID).Exist(&IssueDependency{})
+
+	return
+}
+
+// IssueDependencyIssue custom type for mysql join
+type IssueDependencyIssue struct {
```

Comment:
This is completely unnecessary since you're already tracking parent<->child in `IssueDependecy struct`

Reply:
But i need this to join issue dependencies with issues to check if there are any open dependent issues, the info if the issue is open or closed is saved with the issue, not the dependency.

I have an issue with dependencies. To know if any of these dependent issues is still open, I need to join `issue_dependencies` (relation issue->dependent_issue) with `issue` (informations about the issue, we want to know if it is still open). And to this, I need this struct (or is there a better way to join two tables via xorm?).
See the function `IssueNoDependenciesLeft` later in that file.

Or am I completly wrong?

Reply:
Why do you need a join table? xorm has support for doing JOINs while only working on a single table 🤔 

/cc @lunny

Reply:
Something I did in another project a while back:
```go
contribs := make([]*Contribution, 0, pageSize)
engine.Table("contribution").
  Select("contribution.*, compo.*").
  Join("INNER", "compo", "compo.id = contribution.compo_id").
  Where("contribution.name=?", name).
  Limit(pageSize, (page-1)*pageSize).
  Find(&contribs)
```

Reply:
Oookay, i didn't know that, I put it just like on xorm's official documentation: http://gobook.io/read/github.com/go-xorm/manual-en-US/chapter-05/5.join.html

But the thing you have does indeed look a lot cleaner, I will change it to something like this. Thanks!

Reply:
Done.